### ISSUE TITLE
Narrowing union of TypedDicts with Enum literal does not work with is keyword

### ISSUE THREAD

**Issue by asleep-cult** [microsoft/pyright#3716]
**Describe the bug**
Narrowing a union of TypedDicts with an Enum literal field works with a regular equality check but not the `is` keyword. 

**Expected behavior**
Narrowing should work with both `is` and `==`.

**Screenshots or Code**
```python
import enum
import typing

class FruitType(enum.Enum):
    APPLE = enum.auto()
    BANANA = enum.auto()

class Apple(typing.TypedDict):
    type: typing.Literal[FruitType.APPLE]

class Banana(typing.TypedDict):
    type: typing.Literal[FruitType.BANANA]

Fruit = typing.Union[Apple, Banana]

def eat_fruit(fruit: Fruit) -> None:
    if fruit['type'] is FruitType.APPLE:
        reveal_type(fruit)  # Type of "fruit" is "Apple | Banana"
```

**VS Code extension or command-line**
pyright 1.1.261 command line

**Comment by erictraut** [comment]
Yes, that is not one of the supported type narrowing patterns. For a full list see [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards). As you point out, you can use `==` or `!=` if you want type narrowing to work.

**Comment by erictraut** [comment]
We do occasionally add support for new type narrowing patterns, but each one requires specialized logic and tends to increase the complexity of the type checker and, in some cases, reduce the overall analysis speed. We therefore require good signal from users that it is a frequently-used pattern. In this case, there is an obvious and straightforward alternative pattern that is already supported. We might consider adding support for the alternative `is` and `is not` form in the future if there is sufficient user signal in the form of upvotes. For now, I'm going to close this issue.

**Comment by asleep-cult** [comment]
Oh alright, thank you! On a semi-related note, would it ever be possible to assign `fruit['type']` to a variable and narrow from that value (perhaps if everything were final)?

For example:
```py
def eat_fruit(fruit: Fruit) -> None:
    fruit_type = fruit['type']

    if fruit_type == FruitType.APPLE:
        reveal_type(fruit)  # Type of "fruit" is "Apple"
```

I do realize that this isn't supported for regular objects either but I need to check against quite a few values, so I'd have to write out `fruit['type']` repeatedly which is definitely not optimal.

**Comment by erictraut** [comment]
It's unlikely that we'd ever add support for narrowing that involves intermediate variables like that. It would be very computationally expensive, and the tradeoff wouldn't be worth it for the slow-down.

Pyright does already have support for a feature called [aliased conditional expressions](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#aliased-conditional-expression). It allows you to compute the conditional expression once and assign it to a temporary variable. Perhaps that would help you reduce some of the redundant code.

```python
def eat_fruit(fruit: Fruit) -> None:
    is_apple = fruit["type"] == FruitType.APPLE

    if is_apple:
        reveal_type(fruit)  # Type of "fruit" is "Apple"
```
